Enable dumping only part of the buffer for burst writer.

DomFijan opened the pull request:
## Description
Adds 3 options to dump function of burst writer: start, stop and clear buffer. Start and stop determine which subset of frames will be dumped, while clear buffer flag clears the buffer  or not. Default options should give same results as the current implementation.

## Motivation and context
This would be very useful to have for interactivity with dupin where often we just want to dump a  subset of the deque buffer of dump writer.

## How has this been tested?
Tests testing new functionality have been added.

## Change log

```
Changed:

* Add option to dump a subset of frames in burst writer
  (`#1870 <https://github.com/glotzerlab/hoomd-blue/pull/1870>`__)
```

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/CONTRIBUTING.rst).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/ContributorAgreement.md).
- [x] My name is on the list of contributors (`sphinx-doc/credits.rst`) in the pull request source branch.

DomFijan committed:
add burst writer dump options to dump only subset of stored frames

DomFijan committed:
lint

DomFijan committed:
fix typo

DomFijan committed:
fix constructor

DomFijan committed:
fix if

DomFijan committed:
fix whitespace

DomFijan committed:
add tests

DomFijan committed:
change buffer type

DomFijan committed:
improve error messages

DomFijan committed:
cast to long int

DomFijan committed:
cast to long int

DomFijan committed:
use long int instead of unsigned

DomFijan committed:
use -1 for end for whole deque

DomFijan committed:
fix partial dumping indexing

DomFijan committed:
lint

DomFijan committed:
change error condition

janbridley commented on hoomd/md/pytest/test_burst_writer.py:
```diff
@@ -242,3 +248,36 @@ def test_write_burst_log(sim, tmp_path):
+    burst_writer.flush()
+    if sim.device.communicator.rank == 0:
+        with gsd.hoomd.open(name=filename, mode='r') as traj:
+            assert len(traj) == 6 if empty_buffer else 8
```
```suggestion
            assert len(traj) == (6 if empty_buffer else 8)
```

Prevents the unintended `assert 8`

janbridley committed:
Add missing parentheses

DomFijan committed:
fix test

DomFijan committed:
swap empty buffer with clear entire buffer.
When the whole buffer is not cleared, only part of the buffer up until the last frame that was written is cleared instead

DomFijan committed:
fix logic for partial deletion

DomFijan committed:
lint

DomFijan committed:
add more tests

janbridley commented on hoomd/write/gsd_burst.py:
```diff
@@ -114,20 +114,28 @@ def _attach_hook(self):
 
+        Args:
+            start (int): The first frame to write. Defaults to 0.
+            end (int): The last frame to write, must be positive integer.
```
```suggestion
            end (int): The last frame to write.
```
Default is not a positive int

janbridley commented on hoomd/GSDDequeWriter.cc:
```diff
@@ -59,14 +59,42 @@ void GSDDequeWriter::analyze(uint64_t timestep)
+        throw std::runtime_error("Burst.dump's start index is out of range.");
+        }
+    long int iterator_start, iterator_end;
+    if (end < 0)
+        {
+        iterator_end = buffer_length - start;
+        iterator_start = 0;
+        }
+    else
+        {
+        iterator_end = buffer_length - start;
+        iterator_start = buffer_length - end;
+        }
```
So if `end < 0`, we write out the range `0, length-start`? Why is the correct range not `start, length+end`?

DomFijan replied:
Anything beyond buffer_length would cause out of bounds accessing problems. 
The reason why this is weird is because the dequeue here is in reverse order. So to dump the frames in their time progression we have to do this weird index swapping stuff.

janbridley commented on hoomd/GSDDequeWriter.cc:
```diff
@@ -59,14 +59,42 @@ void GSDDequeWriter::analyze(uint64_t timestep)
+        iterator_end = buffer_length - start;
+        iterator_start = buffer_length - end;
+        }
+    for (auto i = iterator_end - 1; i >= iterator_start; --i)
         {
         write(m_frame_queue[i], m_log_queue[i]);
         }
```
Does this not write the trajectory in reverse? I could be misunderstanding how write works

joaander replied:
See line 54 which is `m_log_queue.push_front(getLogData());`. Why this is `push_front` I do not know. But frames are added at the front of the queue and removed from the end.

DomFijan replied:
It does not. There are tests that specifically test if correct frames are dumped out. The reason why this works is because the dequeue is for some reason ordered in reverse (unsure why)...

joaander commented on hoomd/GSDDequeWriter.h:
```diff
@@ -31,7 +31,7 @@ class PYBIND11_EXPORT GSDDequeWriter : public GSDDumpWriter
     void analyze(uint64_t timestep) override;
 
-    void dump();
+    void dump(long int start, long int end, bool clear_entire_buffer);
```
Let's move `clear_entire_buffer` to a class level attribute (with get/set methods in C++ and a `param_dict` entry in Python). This way, users only need to set the parameter once and leave it.

I think the default could be either `True` or `False`. For backward compatibility, `dump()` with no arguments writes and clears all frames with either setting. We should set the default to be what we think is the most common use-case.

DomFijan committed:
Update hoomd/write/gsd_burst.py

Co-authored-by: Jen Bradley <55467578+janbridley@users.noreply.github.com>

DomFijan committed:
move buffer clearing argument to class constructor

DomFijan committed:
add setter and getter for clear whole buffer after dump

joaander committed:
Document clear_whole_buffer in args.

joaander reviewed (approved):
Thanks!

joaander committed:
pre-commit
